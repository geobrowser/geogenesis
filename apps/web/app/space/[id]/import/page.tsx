// import { Component } from '~/app/space/[id]/import/component';

type ImportPageProps = {
  params: { id: string };
};

export default function ImportPage({ params }: ImportPageProps) {
  return null;

  // return <Component spaceId={params.id} />;
}
