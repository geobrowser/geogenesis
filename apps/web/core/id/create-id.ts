import { createGeoId } from '@geogenesis/sdk';
import { v4, validate, version } from 'uuid';

import { Triple } from '../types';

export function createEntityId() {
  return createGeoId();
}

export function getAppTripleId(triple: Pick<Triple, 'entityId' | 'attributeId'>, spaceId: string) {
  return `${spaceId}:${triple.entityId}:${triple.attributeId}`;
}

/**
 * Triple id encoding should match between client and network.
 * As a future improvement, we could try to run the same code between assemblyscript/typescript.
 */
export function createTripleId(triple: Pick<Triple, 'attributeId' | 'space' | 'entityId'>): string {
  return `${triple.space}:${triple.entityId}:${triple.attributeId}`;
}

export function createValueId() {
  return v4();
}

export const BUILTIN_ENTITY_IDS = ['name', 'type', 'attribute', 'space'];

function isValidUuid(uuid: string) {
  return validate(uuid) && version(uuid) === 4;
}

export function isValidEntityId(id: string) {
  return isValidUuid(id) || BUILTIN_ENTITY_IDS.includes(id);
}
