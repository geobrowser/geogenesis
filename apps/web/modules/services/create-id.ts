import { v4, validate, version } from 'uuid';
import { OmitStrict, Triple } from '../types';

export function createEntityId() {
  return v4();
}

/**
 * Triple id encoding should match between client and network.
 * As a future improvement, we could try to run the same code between assemblyscript/typescript.
 */
export function createTripleId(triple: OmitStrict<Triple, 'attributeName' | 'entityName' | 'id'>): string {
  return `${triple.space}:${triple.entityId}:${triple.attributeId}:${triple.value.id}`;
}

export function createTripleWithId(triple: Triple): Triple {
  return {
    id: createTripleId(triple),
    entityId: triple.entityId,
    attributeId: triple.attributeId,
    attributeName: triple.attributeName,
    value: triple.value,
    space: triple.space,
    entityName: triple.entityName,
  };
}

export function createValueId() {
  return v4();
}

export const BUILTIN_ENTITY_IDS = ['name', 'type', 'attribute', 'space'];

function isValidUuid(uuid: string) {
  return validate(uuid) && version(uuid) === 4;
}

export function isValidEntityId(id: string) {
  return isValidUuid(id) || BUILTIN_ENTITY_IDS.includes(id);
}
