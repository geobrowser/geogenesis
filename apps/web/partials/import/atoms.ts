import { atom } from 'jotai';

export const loadingAtom = atom<boolean>(false);

export const stepAtom = atom<string>('step1');

export const recordsAtom = atom<Array<Array<string>>>([]);

export const headersAtom = atom(get => {
  const records = get(recordsAtom);
  return records?.[0] ?? [];
});

export const examplesAtom = atom(get => {
  const records = get(recordsAtom);
  return records?.[1] ?? [];
});

export const entityCountAtom = atom(get => {
  const records = get(recordsAtom).filter(e => e.length);
  return ((records?.length || 1) - 1).toLocaleString('en-US', { style: 'decimal' });
});

export const entityCountByTypeAtom = atom(get => {
  const actions = get(actionsAtom);

  const typeActions = actions.filter(action => action.attributeId === 'type');

  const entitySetByType: Record<string, Set<string>> = {};

  typeActions.forEach(action => {
    if (action.value.type !== 'entity' || !action.value.name) return;

    if (!Object.hasOwn(entitySetByType, action.value.name)) {
      entitySetByType[action.value.name] = new Set();
    }

    entitySetByType[action.value.name].add(action.entityId);
  });

  const entityCountByType: Array<{ name: string; count: string }> = [];

  Object.keys(entitySetByType).forEach(name => {
    entityCountByType.push({ name, count: entitySetByType[name].size.toLocaleString('en-US', { style: 'decimal' }) });
  });

  return entityCountByType;
});

export const actionsAtom = atom<Array<any>>([]);

export const actionsCountAtom = atom(get => {
  const actions = get(actionsAtom);
  return actions.length.toLocaleString('en-US', { style: 'decimal' });
});

export const publishAtom = atom<boolean>(false);
