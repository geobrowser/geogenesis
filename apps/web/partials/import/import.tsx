'use client';

import type { Space } from '~/core/types';

import { Generate } from './generate';
import { Publish } from './publish';

type ImportProps = {
  spaceId: string;
  space: Space;
};

export const Import = ({ spaceId, space }: ImportProps) => {
  return (
    <>
      <Generate spaceId={spaceId} space={space} />
      <Publish spaceId={spaceId} space={space} />
    </>
  );
};
