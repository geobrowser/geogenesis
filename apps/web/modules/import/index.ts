import { importCSVFile } from './import';

export const Import = {
  importCSVFile,
};
