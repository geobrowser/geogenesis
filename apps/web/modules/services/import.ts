import { parse as parseCSV } from 'papaparse';
import { Triple, Value } from '../types';
import { BUILTIN_ENTITY_IDS, createEntityId, createTripleWithId, isValidEntityId } from './create-id';

export function readFileAsText(file: File) {
  return new Promise<string>(resolve => {
    const reader = new FileReader();
    reader.addEventListener('load', event => {
      const data = event.target!.result;
      resolve(data as string);
    });
    reader.readAsText(file);
  });
}

type CreateUuid = (value: string) => string;

export type EavRow = [string, string, string];

function readCSV(csv: string): EavRow[] {
  const results = parseCSV<string[]>(csv);

  const rows = results.data.slice(1);

  if (rows.some(row => row.length < 3)) {
    throw new Error('Each row must have 3 cells');
  }

  return rows as EavRow[];
}

function unique(triples: Triple[]): Triple[] {
  return Object.values(Object.fromEntries(triples.map(triple => [triple.id, triple])));
}

export function eavRowsToTriples(rows: EavRow[], space: string, createId: CreateUuid = createEntityId): Triple[] {
  // Create a collection of all known entity ids
  const entityIds = new Set([
    ...BUILTIN_ENTITY_IDS,
    ...rows.flatMap(([entityId, attributeId]) => [entityId, attributeId]),
  ]);

  // Ensure entity ids are either valid uuids or builtins
  const entityIdMap = Object.fromEntries([...entityIds].map(id => [id, isValidEntityId(id) ? id : createId(id)]));

  // Create triples, attempting to detect entity references
  const triples = rows.map((row): Triple => {
    const [entityId, attributeId, value] = row;

    const mappedEntityId = entityIdMap[entityId];
    const mappedAttributeId = entityIdMap[attributeId];
    const mappedValue: Value =
      value in entityIdMap
        ? { type: 'entity', id: entityIdMap[value] }
        : { type: 'string', id: createId(value), value };

    return createTripleWithId(space, mappedEntityId, mappedAttributeId, mappedValue);
  });

  return unique(triples);
}

export function convertHealthData(
  csv: string,
  { rowCount = Infinity, shouldIncludeSections = true }: { rowCount?: number; shouldIncludeSections?: boolean } = {
    rowCount: Infinity,
    shouldIncludeSections: true,
  }
): EavRow[] {
  type HealthDataRow = {
    ID_content: string;
    'Title/Fact': string;
    Content: string;
    'Entity ID \n(Source)': string;
    Source: string;
    'Entity ID \n(Guest1)': string;
    'Source\n(Guest1)': string;
    'Entity ID \n(Guest2)': string;
    'Source\n(Guest2)': string;
    'Entity ID \n(Guest3)': string;
    'Source\n(Guest3)': string;
    ID_all_guest: string;
    'Entity ID \n(Location)': string;
    SourceLoc: string;
    'Entity ID \n(Category)': string;
    Category: string;
    'Entity ID \n(Platform)': string;
    'Platform/Site': string;
    Date: string;
    ID_Tag1: string;
    Tag1: string;
    ID_Tag2: string;
    Tag2: string;
    ID_Tag3: string;
    Tag3: string;
    ID_Tag4: string;
    Tag4: string;
    ID_Tag5: string;
    Tag5: string;
    ID_Tag6: string;
    Tag6: string;
    ID_Tag7: string;
    Tag7: string;
    ID_Tag8: string;
    Tag8: string;
    ID_Tag9: string;
    Tag9: string;
    ID_Tag10: string;
    Tag10: string;
    ID_Tag11: string;
    Tag11: string;
    ID_all_tags0: string;
    ID_all_tags: string;
    'UNIQUE TAGS': string;
    Tags: string;
    'All tags': string;
  };

  const results = parseCSV<HealthDataRow>(csv, { header: true });

  const attributeRows: EavRow[] = [
    ['fact', 'type', 'type'],
    ['fact', 'name', 'Fact'],
    ['source', 'type', 'attribute'],
    ['source', 'name', 'Source'],
    ['source location', 'type', 'attribute'],
    ['source location', 'name', 'Source Location'],
    ['category', 'type', 'attribute'],
    ['category', 'name', 'Category'],
    ['sourceType', 'type', 'attribute'],
    ['sourceType', 'name', 'Source Type'],
    ['tag', 'type', 'attribute'],
    ['tag', 'name', 'Tag'],
    ['guest', 'type', 'attribute'],
    ['guest', 'name', 'Guest'],
    ['section', 'type', 'attribute'],
    ['section', 'name', 'Is from Section'],
  ];

  function toEavRow(row: HealthDataRow): EavRow[] {
    const tagTuple = [
      { id: row.ID_Tag1, name: row.Tag1 },
      { id: row.ID_Tag2, name: row.Tag2 },
      { id: row.ID_Tag3, name: row.Tag3 },
      { id: row.ID_Tag4, name: row.Tag4 },
      { id: row.ID_Tag5, name: row.Tag5 },
      { id: row.ID_Tag6, name: row.Tag6 },
      { id: row.ID_Tag7, name: row.Tag7 },
      { id: row.ID_Tag8, name: row.Tag8 },
      { id: row.ID_Tag9, name: row.Tag9 },
      { id: row.ID_Tag10, name: row.Tag10 },
      { id: row.ID_Tag11, name: row.Tag11 },
    ].filter(({ id }) => id !== '');

    const guestTuple = [
      { id: row['Entity ID \n(Guest1)'], name: row['Source\n(Guest1)'] },
      { id: row['Entity ID \n(Guest2)'], name: row['Source\n(Guest2)'] },
      { id: row['Entity ID \n(Guest3)'], name: row['Source\n(Guest3)'] },
    ].filter(({ id }) => id !== '');

    return [
      [row.ID_content, 'type', 'fact'],
      [row.ID_content, 'name', row.Content],
      [row.ID_content, 'source', row['Entity ID \n(Source)']],
      [row.ID_content, 'source location', row['Entity ID \n(Location)']],
      [row.ID_content, 'category', row['Entity ID \n(Category)']],
      [row.ID_content, 'sourceType', row['Entity ID \n(Platform)']],
      [row['Entity ID \n(Source)'], 'name', row.Source],
      [row['Entity ID \n(Location)'], 'name', row.SourceLoc],
      [row['Entity ID \n(Category)'], 'name', row.Category],
      [row['Entity ID \n(Platform)'], 'name', row['Platform/Site']],
      ...tagTuple.map(({ id }): EavRow => [row.ID_content, 'tag', id]),
      ...tagTuple.map(({ id, name }): EavRow => [id, 'name', name]),
      ...guestTuple.map(({ id }): EavRow => [row.ID_content, 'guest', id]),
      ...guestTuple.map(({ id, name }): EavRow => [id, 'name', name]),
    ];
  }

  const chunks = chunkBy(results.data, (rowA, rowB) => {
    if (rowA['Title/Fact'] === 'Title' && rowB['Title/Fact'] === 'Title') return false;
    if (rowA['Title/Fact'] === 'Fact' && rowB['Title/Fact'] === 'Title') return false;
    return true;
  }).filter(chunk => chunk.length > 1); // filter chunks that only contain a title

  const sectionEavs = chunks.flatMap((chunk): EavRow[] => {
    const titleRow = chunk[0];
    const factRows = chunk.slice(1);

    // Add reference to section
    return [
      ...factRows.map((factRow): EavRow => [factRow.ID_content, 'section', titleRow.ID_content]),
      [titleRow.ID_content, 'type', 'section'],
      [titleRow.ID_content, 'name', titleRow.Content],
    ];
  });

  const eavRows = results.data.slice(0, rowCount).flatMap(toEavRow);

  return [...attributeRows, ...(shouldIncludeSections ? sectionEavs : []), ...eavRows];
}

export async function importCSVFile(
  file: File,
  space: string,
  createId: CreateUuid = createEntityId
): Promise<Triple[]> {
  const csv = await readFileAsText(file);
  const rows = file.name === 'healthdata.csv' ? convertHealthData(csv) : readCSV(csv);
  return eavRowsToTriples(rows, space, createId);
}

export function chunkBy<T>(values: T[], belongInSameGroup: (a: T, b: T) => boolean): T[][] {
  if (values.length === 0) return [];

  const result: T[][] = [];

  let start = 0;
  const end = values.length;

  for (let i = start + 1; i < end; i++) {
    const prev = values[i - 1];
    const next = values[i];

    if (!belongInSameGroup(prev, next)) {
      result.push(values.slice(start, i));
      start = i;
    }
  }

  result.push(values.slice(start, end));

  return result;
}
