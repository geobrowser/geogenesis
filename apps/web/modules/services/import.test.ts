import { readFileSync } from 'fs';
import path from 'path';
import { v4 as uuid } from 'uuid';
import { describe, expect, it } from 'vitest';
import { createTripleWithId } from './create-id';
import { convertHealthData, eavRowsToTriples, importCSVFile, readFileAsText, unique } from './import';

function readMockFile(filename: string) {
  const simple = readFileSync(path.join(__dirname, 'mocks', filename), 'utf8');
  return new File([simple], filename);
}

function deterministicUuid(value: string) {
  const random = value
    .slice(0, 16)
    .padEnd(16, '0')
    .split('')
    .map(c => c.charCodeAt(0));

  return uuid({ random });
}

describe('CSV Import', () => {
  it('imports csv', async () => {
    const file = readMockFile('simple.csv');
    const triples = await importCSVFile(file, 's', deterministicUuid);

    expect(triples).toMatchSnapshot();
  });

  it('imports health data format', async () => {
    const file = readMockFile('healthdata.csv');
    const csv = await readFileAsText(file);
    const rows = convertHealthData(csv, { rowCount: 1, shouldIncludeSections: false });

    // So we can see what matches in snapshot tests
    const mockId = (value: string) => value;

    const triples = eavRowsToTriples(rows, 's', mockId);

    expect(triples).toMatchSnapshot();
  });
});

describe('Unique', () => {
  it('returns unique values', () => {
    expect(
      unique([
        createTripleWithId({
          attributeId: 'name',
          entityId: 'e1',
          value: {
            type: 'string',
            value: 'John',
            id: 'v1',
          },
          space: 'space-1',
        }),
        createTripleWithId({
          attributeId: 'name',
          entityId: 'e1',
          value: {
            type: 'string',
            value: 'John',
            id: 'v2',
          },
          space: 'space-1',
        }),
      ])
    ).toEqual([
      {
        id: 'space-1:e1:name:v2',
        attributeId: 'name',
        entityId: 'e1',
        value: {
          type: 'string',
          value: 'John',
          id: 'v2',
        },
        space: 'space-1',
      },
    ]);
  });
});
