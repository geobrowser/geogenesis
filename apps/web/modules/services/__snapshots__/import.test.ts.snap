// Vitest Snapshot v1

exports[`CSV Import > imports csv 1`] = `
[
  {
    "attributeId": "name",
    "entityId": "31303030-3030-4030-b030-303030303030",
    "id": "s:31303030-3030-4030-b030-303030303030:name:44657669-6e30-4030-b030-303030303030",
    "space": "s",
    "value": {
      "id": "44657669-6e30-4030-b030-303030303030",
      "type": "string",
      "value": "Devin",
    },
  },
  {
    "attributeId": "type",
    "entityId": "31303030-3030-4030-b030-303030303030",
    "id": "s:31303030-3030-4030-b030-303030303030:type:70657273-6f6e-4030-b030-303030303030",
    "space": "s",
    "value": {
      "id": "70657273-6f6e-4030-b030-303030303030",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "32303030-3030-4030-b030-303030303030",
    "id": "s:32303030-3030-4030-b030-303030303030:name:53616d30-3030-4030-b030-303030303030",
    "space": "s",
    "value": {
      "id": "53616d30-3030-4030-b030-303030303030",
      "type": "string",
      "value": "Sam",
    },
  },
  {
    "attributeId": "type",
    "entityId": "32303030-3030-4030-b030-303030303030",
    "id": "s:32303030-3030-4030-b030-303030303030:type:70657273-6f6e-4030-b030-303030303030",
    "space": "s",
    "value": {
      "id": "70657273-6f6e-4030-b030-303030303030",
      "type": "entity",
    },
  },
  {
    "attributeId": "type",
    "entityId": "70657273-6f6e-4030-b030-303030303030",
    "id": "s:70657273-6f6e-4030-b030-303030303030:type:type",
    "space": "s",
    "value": {
      "id": "type",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "70657273-6f6e-4030-b030-303030303030",
    "id": "s:70657273-6f6e-4030-b030-303030303030:name:50657273-6f6e-4030-b030-303030303030",
    "space": "s",
    "value": {
      "id": "50657273-6f6e-4030-b030-303030303030",
      "type": "string",
      "value": "Person",
    },
  },
]
`;

exports[`CSV Import > imports health data format 1`] = `
[
  {
    "attributeId": "type",
    "entityId": "fact",
    "id": "s:fact:type:type",
    "space": "s",
    "value": {
      "id": "type",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "fact",
    "id": "s:fact:name:Fact",
    "space": "s",
    "value": {
      "id": "Fact",
      "type": "string",
      "value": "Fact",
    },
  },
  {
    "attributeId": "type",
    "entityId": "source",
    "id": "s:source:type:attribute",
    "space": "s",
    "value": {
      "id": "attribute",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "source",
    "id": "s:source:name:Source",
    "space": "s",
    "value": {
      "id": "Source",
      "type": "string",
      "value": "Source",
    },
  },
  {
    "attributeId": "type",
    "entityId": "source location",
    "id": "s:source location:type:attribute",
    "space": "s",
    "value": {
      "id": "attribute",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "source location",
    "id": "s:source location:name:Source Location",
    "space": "s",
    "value": {
      "id": "Source Location",
      "type": "string",
      "value": "Source Location",
    },
  },
  {
    "attributeId": "type",
    "entityId": "category",
    "id": "s:category:type:attribute",
    "space": "s",
    "value": {
      "id": "attribute",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "category",
    "id": "s:category:name:Category",
    "space": "s",
    "value": {
      "id": "Category",
      "type": "string",
      "value": "Category",
    },
  },
  {
    "attributeId": "type",
    "entityId": "sourceType",
    "id": "s:sourceType:type:attribute",
    "space": "s",
    "value": {
      "id": "attribute",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "sourceType",
    "id": "s:sourceType:name:Source Type",
    "space": "s",
    "value": {
      "id": "Source Type",
      "type": "string",
      "value": "Source Type",
    },
  },
  {
    "attributeId": "type",
    "entityId": "tag",
    "id": "s:tag:type:attribute",
    "space": "s",
    "value": {
      "id": "attribute",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "tag",
    "id": "s:tag:name:Tag",
    "space": "s",
    "value": {
      "id": "Tag",
      "type": "string",
      "value": "Tag",
    },
  },
  {
    "attributeId": "type",
    "entityId": "guest",
    "id": "s:guest:type:attribute",
    "space": "s",
    "value": {
      "id": "attribute",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "guest",
    "id": "s:guest:name:Guest",
    "space": "s",
    "value": {
      "id": "Guest",
      "type": "string",
      "value": "Guest",
    },
  },
  {
    "attributeId": "type",
    "entityId": "section",
    "id": "s:section:type:attribute",
    "space": "s",
    "value": {
      "id": "attribute",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "section",
    "id": "s:section:name:Is from Section",
    "space": "s",
    "value": {
      "id": "Is from Section",
      "type": "string",
      "value": "Is from Section",
    },
  },
  {
    "attributeId": "type",
    "entityId": "10000000",
    "id": "s:10000000:type:fact",
    "space": "s",
    "value": {
      "id": "fact",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "10000000",
    "id": "s:10000000:name:Episode 1: How Your Brain Works & Changes | Huberman Lab",
    "space": "s",
    "value": {
      "id": "Episode 1: How Your Brain Works & Changes | Huberman Lab",
      "type": "string",
      "value": "Episode 1: How Your Brain Works & Changes | Huberman Lab",
    },
  },
  {
    "attributeId": "source",
    "entityId": "10000000",
    "id": "s:10000000:source:1000001",
    "space": "s",
    "value": {
      "id": "1000001",
      "type": "entity",
    },
  },
  {
    "attributeId": "source location",
    "entityId": "10000000",
    "id": "s:10000000:source location:2000133",
    "space": "s",
    "value": {
      "id": "2000133",
      "type": "entity",
    },
  },
  {
    "attributeId": "category",
    "entityId": "10000000",
    "id": "s:10000000:category:3000000",
    "space": "s",
    "value": {
      "id": "3000000",
      "type": "entity",
    },
  },
  {
    "attributeId": "sourceType",
    "entityId": "10000000",
    "id": "s:10000000:sourceType:4000006",
    "space": "s",
    "value": {
      "id": "4000006",
      "type": "entity",
    },
  },
  {
    "attributeId": "name",
    "entityId": "1000001",
    "id": "s:1000001:name:Dr. Andrew Huberman",
    "space": "s",
    "value": {
      "id": "Dr. Andrew Huberman",
      "type": "string",
      "value": "Dr. Andrew Huberman",
    },
  },
  {
    "attributeId": "name",
    "entityId": "2000133",
    "id": "s:2000133:name:Episode 01: How Your Brain Works & Changes | Huberman Lab",
    "space": "s",
    "value": {
      "id": "Episode 01: How Your Brain Works & Changes | Huberman Lab",
      "type": "string",
      "value": "Episode 01: How Your Brain Works & Changes | Huberman Lab",
    },
  },
  {
    "attributeId": "name",
    "entityId": "3000000",
    "id": "s:3000000:name:Health",
    "space": "s",
    "value": {
      "id": "Health",
      "type": "string",
      "value": "Health",
    },
  },
  {
    "attributeId": "name",
    "entityId": "4000006",
    "id": "s:4000006:name:PodcastNotes",
    "space": "s",
    "value": {
      "id": "PodcastNotes",
      "type": "string",
      "value": "PodcastNotes",
    },
  },
]
`;
